lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -42,40 +42,36 @@ struct ConvertComponentOp : public OpConversionPattern<ComponentOp> {
   matchAndRewrite(ComponentOp component, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const override {
     SmallVector<hw::PortInfo> hwInputInfo;
-    for (auto [name, type, direction, _] : component.getPortInfo())
+    auto portInfo = component.getPortInfo();
+    for (auto [name, type, direction, _] : portInfo)
       hwInputInfo.push_back({name, hwDirection(direction), type});
-
-    auto hwMod = rewriter.create<HWModuleOp>(
-        component.getLoc(), component.getNameAttr(), hwInputInfo);
-
-    rewriter.eraseOp(hwMod.getBodyBlock()->getTerminator());
-    ConversionPatternRewriter::InsertionGuard g(rewriter);
-    rewriter.setInsertionPointToEnd(hwMod.getBodyBlock());
+    ModulePortInfo hwPortInfo(hwInputInfo);
 
     SmallVector<Value> argValues;
-    SmallVector<Value> outputWires;
-    size_t portIdx = 0;
-    for (auto [name, type, direction, _] : component.getPortInfo()) {
-      switch (direction) {
-      case calyx::Direction::Input:
-        assert(hwMod.getArgument(portIdx).getType() == type);
-        argValues.push_back(hwMod.getArgument(portIdx));
-        break;
-      case calyx::Direction::Output:
-        auto wire = rewriter.create<sv::WireOp>(component.getLoc(), type, name);
-        auto wireRead =
-            rewriter.create<sv::ReadInOutOp>(component.getLoc(), wire);
-        argValues.push_back(wireRead);
-        outputWires.push_back(wireRead);
-        break;
-      }
-      ++portIdx;
-    }
-
+    auto hwMod = rewriter.create<HWModuleOp>(
+        component.getLoc(), component.getNameAttr(), hwPortInfo,
+        [&](OpBuilder &b, HWModulePortAccessor &ports) {
+          for (auto [name, type, direction, _] : portInfo) {
+            switch (direction) {
+            case calyx::Direction::Input:
+              assert(ports.getInput(name).getType() == type);
+              argValues.push_back(ports.getInput(name));
+              break;
+            case calyx::Direction::Output:
+              auto wire = b.create<sv::WireOp>(component.getLoc(), type, name);
+              auto wireRead =
+                  b.create<sv::ReadInOutOp>(component.getLoc(), wire);
```

Comment:
As I mentioned in https://github.com/llvm/circt/pull/3858, there were a few safety issues that @mikeurbach noticed when looking at this patch. One of them is these two lines which use the `OpBuilder` `b` to create wires. Paraphrasing what he told me, when using a pattern rewriter, you have to remember to use the rewriter to create new ops and not the underlying OpBuilder, since the rewriter needs to keep track of changes.

Locally, I tried changing both of these lines to `rewriter.create()`, but it didn't fix the crashes I was seeing. It sounds like we should still use `rewriter` over `b` here, but we may have other safety issues in this patch.

Reply:
> One of them is these two lines which use the OpBuilder b to create wires. Paraphrasing what he told me, when using a pattern rewriter, you have to remember to use the rewriter to create new ops and not the underlying OpBuilder, since the rewriter needs to keep track of changes.

As far as i can tell, this is already what is being done. When one is using `create` with an `OpBuilder`-derived class, it is that same `OpBuilder` which is passed to the create method, see:
https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Builders.h#L456

Hence changing `b` to `rewriter` should have no effect.

Reply:
I wrote this on the other comment thread, but I think the concern with this specific case is with object slicing. By the time `create` calls the builder, it's just passing an `OpBuilder`: https://github.com/llvm/circt/blob/a610303a74af6d5fa48dc6c4810ab6b9b3ea0f2b/lib/Dialect/HW/HWOps.cpp#L601. That gets used inside the lambda, which means all the stuff on `ConversionPatternRewriter` gets sliced off inside the lambda, so the internal tracking that dialect conversion uses wouldn't work.

I'm not sure this is the root cause of the issue here, but in general it is not valid to mutate IR in a dialect conversion pass via any means other than a ConversionPatternRewriter. That's why in the other comment I also pointed out a situation to use `updateRootInPlace`.

Reply:
Where does the slicing happen? (in the code you've posted it's passed by reference which doesn't cause slicing)

Reply:
Ok the slicing might be a red herring then, because I think builders/rewriters always get passed by reference. That would also explain why switching from `b` to `rewriter` didn't make a difference earlier.